Promote -Wcast-function-type to an error in builds.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46356 Promote -Wcast-function-type to an error in builds.**
* #46227 Remove PyCFunction casts as much as possible.

Adding the flag `-Werror=cast-function-type` to ensure we don't allow
any invalid casts (ex: PyCFunction casts).

For more details see: https://github.com/pytorch/pytorch/issues/45419

Differential Revision: [D24319759](https://our.internmc.facebook.com/intern/diff/D24319759/)

pritamdamania committed:
Promote -Wcast-function-type to an error in builds.

Adding the flag `-Werror=cast-function-type` to ensure we don't allow
any invalid casts (ex: PyCFunction casts).

For more details see: https://github.com/pytorch/pytorch/issues/45419

Differential Revision: [D24319759](https://our.internmc.facebook.com/intern/diff/D24319759/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Promote -Wcast-function-type to an error in builds."


Adding the flag `-Werror=cast-function-type` to ensure we don't allow
any invalid casts (ex: PyCFunction casts).

For more details see: https://github.com/pytorch/pytorch/issues/45419

Differential Revision: [D24319759](https://our.internmc.facebook.com/intern/diff/D24319759/)

[ghstack-poisoned]

albanD reviewed (approved):
Nice!

cc @malfet just FYI and to make sure this is the right change to do in the cmake

pritamdamania committed:
Update on "Promote -Wcast-function-type to an error in builds."


Adding the flag `-Werror=cast-function-type` to ensure we don't allow
any invalid casts (ex: PyCFunction casts).

For more details see: https://github.com/pytorch/pytorch/issues/45419

Differential Revision: [D24319759](https://our.internmc.facebook.com/intern/diff/D24319759/)

[ghstack-poisoned]
